os/fs/driver/mtd/smart.c

```diff
@@ -5109,13 +5147,169 @@ static int smart_joutnal_set_area_state(FAR struct smart_struct_s *dev, int area
+		if (dev->journal_seq < dev->njournaldata / 2) {
+			fdbg("journal sequence & area bug!! sequence : %d\n", dev->journal_seq);
+			return -EFAULT;
+		}
+		psector = dev->neraseblocks * dev->sectorsPerBlk + (dev->njournalsectors / 2);
+		seq = dev->journal_seq - (dev->njournaldata / 2);
+	}
+
+	*offset = (psector * dev->sectorsize) + (seq * sizeof(journal_log_t) + 1);
+	fvdbg("address : %d psector : %d seq : %d dev seq : %d\n", offset, psector, seq, dev->journal_seq);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: smart_journal_checkin
+ *
+ * Description:
+ *   Check in journal data
+ *
+ ****************************************************************************/
+
+static int smart_journal_checkin(FAR struct smart_struct_s *dev, journal_log_t *log, uint32_t address)
+{
+	int ret;
+	journal_log_t buf;
+	size_t log_size = sizeof(journal_log_t);
+
+	/* First write Journal log with Checkin state */
+	ret = MTD_WRITE(dev->mtd, address, log_size, (FAR uint8_t *)log);
+	if (ret != log_size) {
+		fdbg("Checkin Error, write log data... physical address %u ret : %d\n", address, ret);
+		return ret;
+	}
+
+	/* Verify written data */
+	ret = MTD_READ(dev->mtd, address, log_size, (FAR uint8_t *)&buf);
+	if (ret != log_size) {
+		fdbg("Checkin Error, read log data... physical address %u ret : %d\n", address, ret);
+		return ret;
+	}
+
+	if (UINT8TOUINT16(buf.crc16) != smart_calc_journal_crc(&buf)) {
+		fdbg("Checkin Error, verify crc failed buf crc16 : %d calc crc : %d\n", UINT8TOUINT16(buf.crc16), smart_calc_journal_crc(&buf));
+		return ret;
+	}
+
+	/* Increase sequence here, journal_seq is always less than total number of journal data */
+	if (dev->journal_seq == dev->njournaldata -1) {
+		dev->journal_seq = 0;
+	} else {
+		dev->journal_seq++;
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name: smart_journal_checkout
+ *
+ * Description:
+ *   Check in journal data
+ *
+ ****************************************************************************/
+
+static int smart_journal_checkout(FAR struct smart_struct_s *dev, journal_log_t *log, uint32_t address)
+{
+	int ret;
+	uint16_t psector;
+	size_t mtd_size = sizeof(FAR struct smart_sect_header_s);
+	psector = UINT8TOUINT16(log->psector);
+	int type = GET_JOURNAL_TYPE(log->status);
+	fvdbg("type : %d\n", type);
+	
+	/* Update MTD Header on target physical sector First, But Erase doesn't do that */
+	if (type != SMART_JOURNAL_TYPE_ERASE) {
+		ret = MTD_WRITE(dev->mtd, psector * dev->sectorsize, mtd_size, (FAR uint8_t *)&log->mtd_header);
+		if (ret != mtd_size) {
+			fdbg("Checkout error, write mtd header.. ret : %d\n", ret);
+			return ret;
+		}
+	}
+	
+	switch(type) {
+		/* If type is commit, we will check validation of entire sector */
+		case SMART_JOURNAL_TYPE_COMMIT: 
+			ret = MTD_BREAD(dev->mtd, psector * dev->mtdBlksPerSector, dev->mtdBlksPerSector, (FAR uint8_t *)dev->rwbuffer);
```

Comment:
What does MTD_BREAD mean?

Reply:
It is macro to execute mtd_dev_s->bread it read entire of logical sector (physically n * page).
To check crc, we have to read all data in current sector.